### ISSUE TITLE
Orientation issue when resizing images on upload.

### ISSUE THREAD

**Issue by danjames68** [getgrav/grav-plugin-admin#1923]
I've been trying to solve an orientation problem with images uploaded straight from the camera. You can see the whole iteration on the forum here: https://discourse.getgrav.org/t/continuing-image-rotation-issue/12766?u=dan-james

With a fresh grav install I've tracked down 2 issues, and one of them seems to stem from using the Admin Plugin's `Page Media Image Resizer`. If I set it to resize an image, the 'Fix Orientation Automatically' no longer works - it seems to turn the image in the opposite direction.

I tried keeping the metafiles.yaml in the hope they would help but no joy. In addition, if I use `.resize()` after `.fixOrientation` in twig something breaks, but that's separate (I think).

**Comment by pamtbaau** [comment]
It seems the issue is created in the front-end of Admin.

TLDR:
- Uploading image with Exif orientation flag = 6 (means: image requires rotation of **90 degrees clockwise**)
- When resizing is switched on in Admin
  - Resulting uploaded image is being rotated **180 degrees clockwise**
  - Exif remains unchanged: orientation flag = 6

### Test 1: no resizing in Admin
```
pagemedia:
  resize_width: 0
```
Steps:
- front-end
  - Upload image to page (with new name to prevent caching artefacts).
  - The thumbnail in upload field is rotated **180 degrees clockwise**
- back-end
  - At breakpoint in `AdminPlugin::getSubscribedEvents()`
    - Image stored in `$_FILES['file']['tmp_name']`
      - Image data as in original
      - Exif orientation as original
  - Finished upload
    - Image stored in page folder
      - Image data as original
      - Exif orientation as original
- front-end
  - Save page
- backend
  - Hitting breakpoint gregwar Common::resize()
- front-end
  - The thumbnail in upload field is rotated **90 degrees anti-clockwise** (same as original)


### Test 2: with resizing in Admin
```
pagemedia:
  resize_width: 1500
```
Steps:
- front-end
  - Upload image to page (with new name to prevent caching artefacts).
  - The thumbnail in upload field is rotated **180 degrees clockwise**
- back-end
  - At breakpoint in `AdminPlugin::getSubscribedEvents()`
    - Image stored in `$_FILES['file']['tmp_name']`
      - Image data rotated **180 degrees clockwise**
      - Exif orientation as original
  - Finished upload
    - Image stored in page folder
      - Image date rotated **180 degress clockwise**
      - Exif orientation as original
- front-end
  - Save page
- backend
  - Hitting breakpoint gregwar Common::resize()
- front-end
  - The thumbnail in upload field is rotated **180 degrees clockwise**

**Comment by danjames68** [comment]
Thanks Pamtbauu, that's my experience too.

**Comment by ricardo118** [comment]
what version grav and admin is this and could you provide  the images ur using please

**Comment by pamtbaau** [comment]
@ricardo118 Updated my test script with:

Environment:
- Grav v1.6.26
- Admin v1.9.15
- Image from [here](https://dl.dropboxusercontent.com/s/31qijg0s3fp99gj/2020-07-02%2011.42.12.jpg?dl=0)
- Windows 10, WSL 1.0, Ubuntu 20.04

**Comment by rhukster** [comment]
Ok Guys, this is a 2 pronged attempt to resolve the issue.. Please test latest 1.10 branch of admin with 1.7.0 branch of Grav.

First thing, was @w00fz removed the exif library and code we had to solve this before.  For some reason, this was causing it to try to fix things twice!  so an image that was flipped 180 was flipped correctly, then flipped again, to be 360deg and appear broken.  We're not sure how/why it works without the exif.js library as we don't see anything in the changelog to reflect why this would break things all of a sudden.

Next, after we removed this I noticed that the PHP side of things was only handling the fixOrientiation stuff when rendering via markdown. This had the effect of the thumbs being flipped around, but when used in a page it was correct.  However, when used via twig, it was no longer correct! I moved the logic out of `Excerpts` into a lower-level of the `ImageMedium` object.  This appears to have resolved the issues i was seeing with my own test images plus the ones from here: 
https://github.com/recurser/exif-orientation-examples

However, I cannot seem to get the large (8.2mb) Gucci bag example from above (https://dl.dropboxusercontent.com/s/31qijg0s3fp99gj/2020-07-02%2011.42.12.jpg?dl=0) to work. I think this might be due to the large size.

**Comment by pamtbaau** [comment]
TLDR: Compared to previous test:
- No resizing: Same results as previous test.
- With resizing: Images are being rotated **90 degrees clockwise**. In previous test **180 degrees clockwise**.

Installation:
- Installed Grav 1.7.0-rc14
- `$ cd user/plugins/admin`
- `$ rm -rf *`
- `$ git clone -b 1.10 https://github.com/getgrav/grav-plugin-admin.git .`

Used 'Gucci' image downloaded from [here](https://dl.dropboxusercontent.com/s/31qijg0s3fp99gj/2020-07-02%2011.42.12.jpg?dl=0)

### Test 1: no resizing in Admin
```
pagemedia:
  resize_width: 0
```
Steps:
- front-end
  - Upload image to page (with new name to prevent caching artefacts).
  - The thumbnail in upload field is rotated **90 degrees clockwise**
- back-end
  - At breakpoint in `AdminPlugin::getSubscribedEvents()`
    - Image stored in `$_FILES['file']['tmp_name']`
      - Image data as in original
      - Exif orientation as original
  - Finished upload
    - Image stored in page folder
      - Image data as original
      - Exif orientation as original
- front-end
  - Save page
- backend
  - Hitting breakpoint gregwar Common::resize()
- front-end
  - The thumbnail in upload field is rotated **90 degrees anti-clockwise** (same as original)

**Note**: Results are the same as before fix.

### Test 2: with resizing in Admin
```
pagemedia:
  resize_width: 1500
```
Steps:
- front-end
  - Upload image to page (with new name to prevent caching artefacts).
  - The thumbnail in upload field is rotated **90 degrees clockwise**
- back-end
  - At breakpoint in `AdminPlugin::getSubscribedEvents()`
    - Image stored in `$_FILES['file']['tmp_name']`
      - Image data rotated **90 degrees clockwise**
      - Exif orientation as original
  - Finished upload
    - Image stored in page folder
      - Image data rotated **90 degrees clockwise**
      - Exif orientation as original
- front-end
  - Save page
- backend
  - Hitting breakpoint gregwar Common::resize()
- front-end
  - The thumbnail in upload field is rotated **90 degrees clockwise**

**Note**: Images are now rotated **90 degrees clockwise**. Previous test was **180 degrees clockwise**

**Comment by rhukster** [comment]
i appreciate the thorough testing but please don’t use the Gucci image.  As i stated there is something wrong with that image that is causing issues.  please test with the images i linked to in the github repo.  those test all 8 rotation and flip combinations.

**Comment by danjames68** [comment]
Thanks Andy and PamtBauu for looking into this.

I’m going to ask the client for their camera model and investigate if their EXIF has any known issues, just to check if there’s a reason the Gucci image is being problematic.


Dan

**Comment by pamtbaau** [comment]
@rhukster Ran the test using image [Portrait_6.jpg](https://github.com/recurser/exif-orientation-examples/blob/master/Portrait_6.jpg), which has the same orientation (portrait) and rotation (90 degrees anti-clockwise) as the Gucci image.

**Results**: Both images are handled the same. The image data has been rotated **90 degrees clockwise**, and Exif remains the same as original.

Also:
- Both images show same Exif in online Exif extractor http://metapicz.com
- Both images show the same Exif in Gimp.
- Both images are correctly displayed in Windows Explorer
- Both images are correctly displayed by Chrome
- Both images are correctly displayed by Microsoft Photo app
- Both images are correctly displayed by Microsoft Paint
- The above makes me doubt if it is a problem with the Gucci image...

**Comment by rhukster** [comment]
I'm thinking you didn't update the admin to latest 1.10 branch? that's not the behavior I'm seeing.

**Comment by pamtbaau** [comment]
How do you check if image is correctly processed?

I created the following installation:
- Installed Grav 1.7.0-rc14
- `$ cd user/plugins/admin`
- `$ rm -rf *`
- `$ git clone -b 1.10 https://github.com/getgrav/grav-plugin-admin.git .`

`/user/plugins/admin/CHANGELOG.md` contains:

> # v1.10.0-rc.15
> ## mm/dd/2020
>
> 1. [](#bugfix)
>    * Disabled the EXIF library for Dropzone for fixing the orientation as it was getting applied twice [#1923](https://github.com/getgrav/grav-plugin-admin/issues/1923)

**Comment by rhukster** [comment]
that should be correct.  make sure your cache is cleared to pick up the latest JS then. 

This is what i get when i upload. These are the canvas generated thumbnails created by dropzone.  All look correct:

![Image Orientation #1923 | Grav 2020-07-21 14-09-31](https://user-images.githubusercontent.com/1084697/88101858-e0129c80-cb5b-11ea-97ee-97d423a117a0.jpg)

Then when I click 'save', these are the ones generated by Grav via the `listPageMedia` call:

![Image Orientation #1923 | Grav 2020-07-21 14-06-13](https://user-images.githubusercontent.com/1084697/88101564-68dd0880-cb5b-11ea-854a-d46ecccd5568.jpg)

When i drag the thumb into the editor either raw or with cropzoom operation, and click 'preview', it appears correct in the browser. Even the Gucci one:

![Image Orientation #1923 | Grav 2020-07-21 14-11-15](https://user-images.githubusercontent.com/1084697/88101989-17814900-cb5c-11ea-866c-854e62874da7.jpg)

**Comment by rhukster** [comment]
BTW admin image resize is definitely causing the thumbs to be broken, will look a that after. For now i want to find out why the Gucci image is not working fully.

**Comment by rhukster** [comment]
oh btw, make sure you have this on in system -> media:

![Configuration: System | Grav 2020-07-21 14-31-33](https://user-images.githubusercontent.com/1084697/88103931-ee15ec80-cb5e-11ea-8175-3625cf1a1d30.jpg)

**Comment by rhukster** [comment]
Ok, I committed another Grav fix that addresses the specific issue with the Gucci image. That one was related to a file stream that was causing the EXIF to come back as invalid.  This is sorted now.

The other issue for resizing is due to a bug that exists in DropZone code specifically related to resizing of images on upload and EXIF orientations.  We had to fork Dropzone for now to address this, and we also created a PR, so hopefully that will get fixed down the line. https://github.com/enyo/dropzone/pull/1876

I've set the "Fix orientation automatically" to **Yes** by default for 1.7, but this will need to be set on your installs... Other than that, this combination of fixes in Grav + Admin will be released very soon as RC15.

**Comment by pamtbaau** [comment]
> How do you check if image is correctly processed?

- Please be advised that, [since recently](https://www.fxsitecompat.dev/en-CA/docs/2020/jpeg-images-are-now-rotated-by-default-according-to-exif-data/), browsers (Firefox, Chrome) also "fix" the orientation based on the Exif in the image. This may add some confusion.
- The display of the images in Admin are not a good representation of their correctness.

For example:
- Using the same installation as before
- With resizing switched on
  ```
  pagemedia:
    resize_width: 1500
  ```
- With auto fix orientation switched off
  ```
  images:
    auto_fix_orientation: false
  ```

Results in Admin after clicking 'Save': All looks well
![image](https://user-images.githubusercontent.com/5120011/88126785-91ecb000-cbd2-11ea-8667-3c22506f6350.png)

When hitting the 'preview' button next to the image in Admin, the browser shows the following:
![image](https://user-images.githubusercontent.com/5120011/88127172-5b636500-cbd3-11ea-8f69-f3ed9957979a.png)

Windows Explorer shows:
![image](https://user-images.githubusercontent.com/5120011/88127702-9d40db00-cbd4-11ea-9e93-a4c88541cc02.png)

When opening the saved file in Gimp, it shows the data has been rotated, but the Exif is not corrected.
![image](https://user-images.githubusercontent.com/5120011/88127395-c7de6400-cbd3-11ea-82ac-d1a2cda9b173.png)

## With resizing switched off:
- With resizing switched off
  ```
  pagemedia:
    resize_width: 0
  ```
The image are shown as follows in Admin after clicking 'Save':
![image](https://user-images.githubusercontent.com/5120011/88128336-02490080-cbd6-11ea-9629-36d29bc0bcd6.png)

**Comment by rhukster** [comment]
well i don't know what to say at this point. I am not having the issues with latest Grav + Admin as long as i have auto_fix enabled with both regular and resized images.  The images are correct in canvas (dropzone upload), page media list (on save), in the filesystem, as well as in the browser.

In fact i worked on this all afternoon with Djamil and after much digging and testing, we came up with this solution that works for both of us in all scenarios.

I'm at a loss that you are still having this issue. I'll be releasing this version of stuff tomorrow and perhaps we can get some wider testing.

**Comment by rhukster** [comment]
Do you have php-exif support enabled btw? that's the only thing i can think of at this point.

Also check your Grav logs, as I enabled some error handling during `fixOrientation()` that previously silently failed.

**Comment by pamtbaau** [comment]
My latest test are with the previous download of Admin and Grav 1.7 and are intended only to show that the display in Admin is not representative for the images shown in the browser or other apps.

I am in the progress of cloning the latest Admin 1.10 and Grav 1.7 repos and redo the tests.

**Comment by pamtbaau** [comment]
TLDR: All seems to be working well with the latest clones of Admin (branch 1.10) and Grav (branch 1.7)

New installation:
- `$ mkdir site-orientation2`
- `$ cd site-orientation2`
- `$ git clone -b 1.7 https://github.com/getgrav/grav.git .`
- `$ bin/grav install`
- `$ git clone -b 1.10 https://github.com/getgrav/grav-plugin-admin.git .`
- `$ bin/gpm install login`

Test:
- Resize switched on:
  ```
  pagemedia:
    resize_width: 1500
  ```
- Fix orientation switched on as per the default:
  ```
  images:
    auto_fix_orientation: true                     # Automatically fix the image orientation based on the Exif data
  ```

Results in Admin after clicking 'Save': All looks well
![image](https://user-images.githubusercontent.com/5120011/88132638-62dd3b00-cbe0-11ea-981e-0f5d39334856.png)

When hitting the 'preview' button next to the image in Admin, the browser shows the following:
![image](https://user-images.githubusercontent.com/5120011/88132750-a041c880-cbe0-11ea-9de1-2e6a125f0cc7.png)

Windows Explorer shows:
![image](https://user-images.githubusercontent.com/5120011/88132835-c9625900-cbe0-11ea-8377-46b2d7d73f95.png)

When opening the saved file in Gimp, it shows the orientation metatag has been removed.

**Comment by rhukster** [comment]
Phew!!! Glad it's now working as expected.

**Comment by pamtbaau** [comment]
@danjames68 Would you mind installing the latest Grav 1,7 and Admin 1.10 as soon as they are released and test for your use-case and environment? If you're eager, you might try and clone their latest branches 1.7 and 1.10 as I described above.

Note: You won't need the `fixOrientation` function in Twig anymore, since it is set as default behavour in `system.yaml`.

**Comment by danjames68** [comment]
Hi Andy and hi Pamtbaau - thanks so much for this. I've installed latest grav and cloned the Admin 1.1 and it all works as expected. I really appreciate the time everybody put in - I'll be putting some money in the Open collective box...
